src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -134,150 +191,196 @@ public function handle(AddCartRuleToOrderCommand $command): void
-                $orderCartRule->value_tax_excl = $cartRule['value_tax_excl'];
-                $result &= $orderCartRule->add();
+    /**
+     * @param array $cartRules
+     * @param Order $order
+     * @param OrderInvoice|null $orderInvoice
+     */
+    private function buildFreeShippingTypeCartRules(array &$cartRules, Order $order, ?OrderInvoice $orderInvoice): void
+    {
+        if ($orderInvoice) {
+            if ($orderInvoice->total_shipping_tax_incl > 0) {
+                $cartRules[$orderInvoice->id]['value_tax_incl'] = $orderInvoice->total_shipping_tax_incl;
+                $cartRules[$orderInvoice->id]['value_tax_excl'] = $orderInvoice->total_shipping_tax_excl;
 
-                $order->total_discounts += $orderCartRule->value;
-                $order->total_discounts_tax_incl += $orderCartRule->value;
-                $order->total_discounts_tax_excl += $orderCartRule->value_tax_excl;
-                $order->total_paid -= $orderCartRule->value;
-                $order->total_paid_tax_incl -= $orderCartRule->value;
-                $order->total_paid_tax_excl -= $orderCartRule->value_tax_excl;
+                // Update OrderInvoice
+                $this->applyDiscountOnInvoice(
+                    $orderInvoice,
+                    $cartRules[$orderInvoice->id]['value_tax_incl'],
+                    $cartRules[$orderInvoice->id]['value_tax_excl']
+                );
             }
+        } elseif ($order->hasInvoice()) {
+            $orderInvoices = $order->getInvoicesCollection();
+            foreach ($orderInvoices as $orderInvoice) {
+                /** @var OrderInvoice $orderInvoice */
+                if ($orderInvoice->total_shipping_tax_incl <= 0) {
+                    continue;
+                }
+                $cartRules[$orderInvoice->id]['value_tax_incl'] = $orderInvoice->total_shipping_tax_incl;
+                $cartRules[$orderInvoice->id]['value_tax_excl'] = $orderInvoice->total_shipping_tax_excl;
 
-            // Update Order
-            $result &= $order->update();
+                // Update OrderInvoice
+                $this->applyDiscountOnInvoice(
+                    $orderInvoice,
+                    $cartRules[$orderInvoice->id]['value_tax_incl'],
+                    $cartRules[$orderInvoice->id]['value_tax_excl']
+                );
+            }
+        } else {
+            $cartRules[0]['value_tax_incl'] = $order->total_shipping_tax_incl;
+            $cartRules[0]['value_tax_excl'] = $order->total_shipping_tax_excl;
         }
+    }
 
-        if (!$result) {
-            throw new OrderException('An error occurred during the OrderCartRule creation');
+    /**
+     * @param AddCartRuleToOrderCommand $command
+     * @param Order $order
+     * @param array $cartRule
+     * @param float $discountValue
+     *
+     * @return CartRule
+     */
+    private function createCartRule(
+        AddCartRuleToOrderCommand $command,
```

Comment:
You can simply use the cart rule type as a string instead of passing around the command all the time